File: src/main/java/azneotec/algorithms/BinarySearch.java
package azneotec.algorithms;

import java.util.Arrays;

public class BinarySearch {

    /**
     In Java, object1.compareTo(object2)
     -1  -> object1 value < object2 value
     0   -> object1 value = object2 value
     1   -> object1 value > object2 value
     **/
    public static int search(Comparable[] A, Comparable target) {

        int lo = 0;
        int hi = A.length - 1;

        while (lo <= hi) {
            int mid = (lo + hi) / 2;
            int comparisonResult = A[mid].compareTo(target);

            if (comparisonResult == 0)      return mid;
            else if (comparisonResult < 0) lo = mid + 1;
            else                            hi = mid - 1;
        }

        return -1;
    }

    /**
         BONUS INTERVIEW QUESTION: Sorted Shifted Array with Binary search

         GIVEN: A = [3, 4, 5, 1, 2], find the the target: 2
         Runtime Complexity: O(logN) where N is length of Array A

     */

    public static int shiftedBinarySearch(int[] A, int target) {
        int lo = 0;
        int hi = A.length - 1;

        while (lo <= hi) {
            int mid = (lo + hi) / 2;
            int midElement = A[mid];

            if (target == midElement) return mid;

            int leftNum = A[lo];
            int rightNum = A[hi];

            if (leftNum <= midElement) {
                if (target < midElement && target >= leftNum) {
                    hi = mid - 1;
                } else {
                    lo = mid + 1;
                }
            } else {
                if (target > midElement && target <= rightNum) {
                    lo = mid + 1;
                } else {
                    hi = mid - 1;
                }
            }

        }

        return -1;

    }

    public static void main(String[] args) {
//        Integer[] integers = {1,2,3,4,5,6};
//        System.out.println(search(integers, 3));
//        System.out.println(search(integers, 7));
//
//        String[] strings = {"-1", "1","2","3","4","5","6"};
//        System.out.println(search(strings, "-1"));
//        System.out.println(search(strings, "5"));
//        System.out.println(search(strings, "-10"));

        int[] A = {3, 4, 5, 6, 1, 2};
        System.out.println(Arrays.toString(A));
        System.out.println(shiftedBinarySearch(A, 2));
        System.out.println(shiftedBinarySearch(A, 6));
        System.out.println(shiftedBinarySearch(A, 7));
    }

}

File: src/main/scala/algorithms/BinarySearch.scala
package algorithms

import scala.math.Ordering

object BinarySearch extends App {

  def binarySearch[Type](a: Array[Type], target: Type)
                        (implicit ord: Ordering[Type]): Int = {

    var lo = 0
    var hi = a.length - 1

    while(lo <= hi) {
      val mid = (lo + hi) / 2

      val comparisonResult = ord.compare(a(mid), target)

      if (comparisonResult == 0) return mid
      else if (comparisonResult < 0) lo = mid + 1
      else hi = mid - 1

    }

    -1
  }


  def binarySearchVersion2(a: Array[Int], key: Int): Int = {

    var lo = 0
    var hi = a.length - 1

    while (lo <= hi) {
      val mid: Int = lo + (hi - lo) / 2

      val comparisonResult = a(mid).compareTo(key)

      if      (comparisonResult == 0) return mid
      else if (comparisonResult < 0)  lo = mid + 1
      else                            hi = mid - 1
    }

    -1
  }

  val ints: Array[Int] = Array(1, 2, 3, 4, 5, 6)
  println(s"Search Index of 2 = ${binarySearch(ints, 2)}")
  println(s"Search Index of 5 = ${binarySearch(ints, 5)}")
  println(s"Search Index of 10 = ${binarySearch(ints, 10)}")

  val strings: Array[String] = Array("1", "2", "3", "4", "5", "6")
  println(s"Search Index of '2' = ${binarySearch(strings, "2")}")
  println(s"Search Index of '5' = ${binarySearch(strings, "5")}")
  println(s"Search Index of '-21' = ${binarySearch(strings, "-21")}")
}
